File: src/main/并查集/t947/RemoveStones.java
package 并查集.t947;

import java.util.HashSet;

/**
 * @auther: sotowang
 * @date: 2020/2/4 15:12
 */
public class RemoveStones {
    public class UnionFind{
        private int[] parent;
        private int[] size;

        public UnionFind(int n) {
            this.parent = new int[n];
            this.size = new int[n];
            for (int i = 0; i < parent.length; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        public void union(int i, int j) {
            int parentI = find(i);
            int parentJ = find(j);
            if (parentI == parentJ) {
                return;
            }
            if (size[parentI] > size[parentJ]) {
                parent[parentI] = parentJ;
                size[parentJ] += size[parentI];
            } else {
                parent[parentJ] = parentI;
                size[parentI] += size[parentJ];
            }
        }

        private int find(int x) {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            return x;
        }
    }
    public int removeStones(int[][] stones) {
        UnionFind uf = new UnionFind(20000);
        HashSet<Integer> rootSet = new HashSet<>();
        for (int i = 0; i < stones.length; i++) {
            uf.union(stones[i][0], stones[i][1] + 10000);
        }
        for (int i = 0; i < stones.length; i++) {
            rootSet.add(uf.find(stones[i][0]));
        }
        return stones.length - rootSet.size();
    }

    public static void main(String[] args) {
        int[][] stones2 = {{0, 0}};
        assert new RemoveStones().removeStones(stones2) == 0;
        int[][] stones1 = {{0, 0}, {0, 2}, {1, 1}, {2, 0}, {2, 2}};
        assert new RemoveStones().removeStones(stones1) == 3;
        int[][] stones = {{0, 0}, {0, 1}, {1, 0}, {1, 2}, {2, 1}, {2, 2}};
        assert new RemoveStones().removeStones(stones) == 5;

    }
}

File: src/main/并查集/t947/removeStones.go
package t947

type unionFind struct {
	parent []int
	size   []int
}

func generate(n int) *unionFind {
	uf := unionFind{make([]int, n), make([]int, n)}
	for i := 0; i < n; i++ {
		uf.parent[i] = i
		uf.size[i] = 1
	}
	return &uf
}
func (uf *unionFind) union(i, j int) {
	parentI, parentJ := uf.find(i), uf.find(j)
	if parentI == parentJ {
		return
	}
	if uf.size[parentI] > uf.size[parentJ] {
		uf.parent[parentI] = parentJ
		uf.size[parentJ] += uf.size[parentI]
	} else {
		uf.parent[parentJ] = parentI
		uf.size[parentI] += uf.size[parentJ]
	}
}
func (uf *unionFind) find(x int) int {
	for uf.parent[x] != x {
		uf.parent[x] = uf.parent[uf.parent[x]]
		x = uf.parent[x]
	}
	return x
}
func removeStones(stones [][]int) int {
	uf := generate(20000)
	rootSet := make(map[int]bool)
	for i := range stones {
		uf.union(stones[i][0], stones[i][1]+10000)
	}
	for i := range stones {
		x := uf.find(stones[i][0])
		if _, ok := rootSet[x]; !ok {
			rootSet[x] = true
		}
	}
	return len(stones) - len(rootSet)
}
